File: arnav_customization/public/js/stock_entry.js
frappe.ui.form.on("Stock Entry Detail", {

    custom_sku: function(frm, cdt, cdn) {

        let row = locals[cdt][cdn];

        if (!row.custom_sku) return;

        frappe.call({
            method: "arnav_customization.sku_mapping_backend.sku_mapper.get_sku_data",
            args: {
                sku: row.custom_sku
            },

            callback: function(r) {

                if (!r.message) return;

                let d = r.message;

                frappe.model.set_value(cdt, cdn, "item_code", d.item_code);

                // UI display qty as gross weight
                frappe.model.set_value(cdt, cdn, "qty", d.gross_weight);

                // preserve separately
                frappe.model.set_value(cdt, cdn, "custom_gross_weight", d.gross_weight);

                frappe.model.set_value(cdt, cdn, "batch_no", d.batch_no);

                setTimeout(() => {

                    frappe.model.set_value(cdt, cdn, "basic_rate", d.cost_price);

                    frappe.model.set_value(cdt, cdn, "valuation_rate", d.valuation_rate);

                }, 300);

                if (!row.s_warehouse && !row.t_warehouse) {

                    frappe.model.set_value(cdt, cdn, "t_warehouse", d.warehouse);
                }
            }
        });
    }
});


frappe.ui.form.on("Stock Entry", {

    refresh(frm) {

        // enable only for transfer / issue
        if (
            frm.doc.purpose === "Material Transfer" ||
            frm.doc.purpose === "Material Issue"
        ) {

            frm.set_value("custom_use_qty_mode", 1);

        } else {

            frm.set_value("custom_use_qty_mode", 0);
        }
    }
});

File: arnav_customization/sku_mapping_backend/handlers/stock_entry.py
import frappe
from arnav_customization.sku_mapping_backend.sku_service import get_sku_data


def process(doc, method):

    for row in doc.items:

        if not row.custom_sku:
            continue

        sku = get_sku_data(row.custom_sku)

        row.item_code = sku.product

        # DEFAULT UI DISPLAY
        # show gross weight in qty initially
        row.qty = float(sku.gross_weight or 0)

        # preserve gross weight separately
        row.custom_gross_weight = float(sku.gross_weight or 0)

        # IMPORTANT
        row.serial_and_batch_bundle = None

        row.batch_no = sku.batch_no or row.custom_sku

        row.basic_rate = sku.cost_price
        row.valuation_rate = sku.valuation_rate

        if not row.s_warehouse and not row.t_warehouse:
            row.t_warehouse = sku.warehouse


def material_transfer_qty_handler(doc, method):

    # only for transfer / issue
    if doc.purpose not in ["Material Transfer", "Material Issue"]:
        return

    # checkbox disabled = normal ERP behavior
    if not doc.custom_use_qty_mode:
        return

    for row in doc.items:

        # skip invalid rows
        if not row.item_code:
            continue

        # preserve gross weight before switching
        if row.qty and not row.custom_gross_weight:
            row.custom_gross_weight = row.qty

        # actual stock movement qty
        actual_qty = float(row.custom_gross_weight or 1)

        # IMPORTANT:
        # ERP stock movement fields
        row.qty = actual_qty
        row.transfer_qty = actual_qty

        # prevent bundle validation issue
        row.serial_and_batch_bundle = None


# # def material_transfer_qty_handler(doc, method):

# #     # only for material transfer
# #     if doc.purpose != "Material Transfer" or doc.purpose != "Material Issue":
# #         return

# #     # dynamic switch
# #     if not doc.custom_use_qty_mode:
# #         return

# #     for row in doc.items:

# #         # skip empty rows
# #         if not row.custom_gross_weight:
# #             continue

# #         # preserve original gross weight
# #         row.custom_weight = row.qty

# #         # IMPORTANT:
# #         # actual stock qty becomes custom qty
# #         row.qty = float(row.custom_gross_weight)

# #         # avoid transfer qty mismatch
# #         row.transfer_qty = float(row.custom_gross_weight)

